src/articles/slices/articles.ts

```diff
@@ -0,0 +1,137 @@
+import { DateTime } from 'luxon';
+
+import { createAsyncThunk, createEntityAdapter, createSlice, SerializedError } from '@reduxjs/toolkit';
+import { getArticle, getArticles } from 'articles/api';
+import { IArticle } from 'articles/models/Article';
+import { State } from 'core/redux/Store';
+
+const articlesAdapter = createEntityAdapter<IArticle>({
+  sortComparer: (articleA, articleB) => {
+    return Number(DateTime.fromISO(articleA.published_date) > DateTime.fromISO(articleB.published_date));
+  },
+});
+
+export const articleSelectors = articlesAdapter.getSelectors<State>((state) => state.articles);
+
+export const fetchArticlesList = createAsyncThunk('articles/fetchList', async () => {
+  const articles = await getArticles();
+  return articles;
+});
+
+export const fetchArticleById = createAsyncThunk('articles/fetchById', async (articleId: number) => {
+  const articles = await getArticle(articleId);
+  return articles;
+});
+
+export const fetchRelatedArticles = createAsyncThunk(
+  'articles/fetchRelated',
+  async (articleId: number, { getState }) => {
+    const state = getState() as State;
+    const mainArticle = articleSelectors.selectById(state, articleId);
+    if (mainArticle) {
+      const relatedArticles = await Promise.all(mainArticle.tags.map((tag) => getArticles({ tags: tag })));
+      return relatedArticles.flat();
+    }
+    return [];
+  }
+);
+
+export const fetchFrontPageArticles = createAsyncThunk('articles/fetchFrontPage', async () => {
+  const articles = await getArticles({ page_size: 4 });
+  return articles;
+});
+
+interface IState {
+  loading: 'idle' | 'pending';
+  error: SerializedError | null;
+  entities: Record<number, IArticle>;
+  frontPageArticleIds: number[];
+}
+
+const INITIAL_STATE: IState = {
+  loading: 'idle',
+  error: null,
+  entities: {},
+  frontPageArticleIds: [],
+};
+
+export const articlesSlice = createSlice({
+  name: 'articles',
+  initialState: articlesAdapter.getInitialState(INITIAL_STATE),
+  reducers: {},
+  extraReducers: (builder) => {
+    builder.addCase(fetchArticlesList.pending, (state) => {
+      if (state.loading === 'idle') {
+        state.loading = 'pending';
```

Comment:
What is the thought behind this explicit switch? 
If it is to prevent unnecessary rerendering, does Redux have some builtin prevention of that?

Reply:
I can agree the the builder might not be the most readable thing, but I think its better than the alternative. 

Regular actions are _super easy_ with redux toolkit, and we'll get to those later when we actually change our state other than just data loading.
Simple example:
``` typescript
const counterSlice = createSlice({
  name: 'counter',
  initialState: 0,
  reducers: {
    increment: (state) => state + 1,
    decrement: (state) => state - 1,
  },
})

const { increment, decrement } = counterSlice.actions;

// Then you can simply use that in a component like so
const IncrementButton = () => {
  const dispatch = useDispatch();
  
  const onClick = () => {
    dispatch(increment())
  };

  return (
    <button onClick={onClick}>Add to counter</button>
  );
};
```

With thunks the story is a little different, and you can look at how it is done here: https://redux-toolkit.js.org/api/createAsyncThunk, which is almost exactly like the builder. This is because the actions are define outside the reducer. The difference is that it doesn't work as well with typescript, so we have to use the builder :slightly_frowning_face: 
The nice thing about the builder is that the return types of the thunk are inferred automatically :smile:

Reply:
I see! Do you know if it will rerender every time if it is not included?

Reply:
It seems i picked it from a more customized version of their examples which handled 'only requesting one at a time' in a specific way which we are not. I'll make this more general :smile:

Reply:
Neat!